# Library

In [202]:
import pandas as pd
import numpy as np

# Configuration

In [271]:
pd.set_option('display.max_columns', None)
pd.set_option('display.max_colwidth', None)

# Import Dataset

In [204]:
sales = pd.read_csv('dataset/amazon.csv')
sales.head(5)

product_id                                       product_name  \
0  B07JW9H4J1  Wayona Nylon Braided USB to Lightning Fast Cha...   
1  B098NS6PVG  Ambrane Unbreakable 60W / 3A Fast Charging 1.5...   
2  B096MSW6CT  Sounce Fast Phone Charging Cable & Data Sync U...   
3  B08HDJ86NZ  boAt Deuce USB 300 2 in 1 Type-C & Micro USB S...   
4  B08CF3B7N1  Portronics Konnect L 1.2M Fast Charging 3A 8 P...   

                                            category discounted_price  \
0  Computers&Accessories|Accessories&Peripherals|...             ₹399   
1  Computers&Accessories|Accessories&Peripherals|...             ₹199   
2  Computers&Accessories|Accessories&Peripherals|...             ₹199   
3  Computers&Accessories|Accessories&Peripherals|...             ₹329   
4  Computers&Accessories|Accessories&Peripherals|...             ₹154   

  actual_price discount_percentage rating rating_count  \
0       ₹1,099                 64%    4.2       24,269   
1         ₹349                 43%    4.0       43,994   
2       ₹1,899                 90%    3.9        7,928   
3         ₹699                 53%    4.2       94,363   
4         ₹399                 61%    4.2       16,905   

                                       about_product  \
0  High Compatibility : Compatible With iPhone 12...   
1  Compatible with all Type C enabled devices, be...   
2  【 Fast Charger& Data Sync】-With built-in safet...   
3  The boAt Deuce USB 300 2 in 1 cable is compati...   
4  [CHARGE & SYNC FUNCTION]- This cable comes wit...   

                                             user_id  \
0  AG3D6O4STAQKAY2UVGEUV46KN35Q,AHMY5CWJMMK5BJRBB...   
1  AECPFYFQVRUWC3KGNLJIOREFP5LQ,AGYYVPDD7YG7FYNBX...   
2  AGU3BBQ2V2DDAMOAKGFAWDDQ6QHA,AESFLDV2PT363T2AQ...   
3  AEWAZDZZJLQUYVOVGBEUKSLXHQ5A,AG5HTSFRRE6NL3M5S...   
4  AE3Q6KSUK5P75D5HFYHCRAOLODSA,AFUGIFH5ZAFXRDSZH...   

                                           user_name  \
0  Manav,Adarsh gupta,Sundeep,S.Sayeed Ahmed,jasp...   
1  ArdKn,Nirbhay kumar,Sagar Viswanathan,Asp,Plac...   
2  Kunal,Himanshu,viswanath,sai niharka,saqib mal...   
3  Omkar dhale,JD,HEMALATHA,Ajwadh a.,amar singh ...   
4  rahuls6099,Swasat Borah,Ajay Wadke,Pranali,RVK...   

                                           review_id  \
0  R3HXWT0LRP0NMF,R2AJM3LFTLZHFO,R6AQJGUP6P86,R1K...   
1  RGIQEG07R9HS2,R1SMWZQ86XIN8U,R2J3Y1WL29GWDE,RY...   
2  R3J3EQQ9TZI5ZJ,R3E7WBGK7ID0KV,RWU79XKQ6I1QF,R2...   
3  R3EEUZKKK9J36I,R3HJVYCLYOY554,REDECAZ7AMPQC,R1...   
4  R1BP4L2HH9TFUP,R16PVJEXKV6QZS,R2UPDB81N66T4P,R...   

                                        review_title  \
0  Satisfied,Charging is really fast,Value for mo...   
1  A Good Braided Cable for Your Type C Device,Go...   
2  Good speed for earlier versions,Good Product,W...   
3  Good product,Good one,Nice,Really nice product...   
4  As good as original,Decent,Good one for second...   

                                      review_content  \
0  Looks durable Charging is fine tooNo complains...   
1  I ordered this cable to connect my phone to An...   
2  Not quite durable and sturdy,https://m.media-a...   
3  Good product,long wire,Charges good,Nice,I bou...   
4  Bought this instead of original apple, does th...   

                                            img_link  \
0  https://m.media-amazon.com/images/W/WEBP_40237...   
1  https://m.media-amazon.com/images/W/WEBP_40237...   
2  https://m.media-amazon.com/images/W/WEBP_40237...   
3  https://m.media-amazon.com/images/I/41V5FtEWPk...   
4  https://m.media-amazon.com/images/W/WEBP_40237...   

                                        product_link  
0  https://www.amazon.in/Wayona-Braided-WN3LG1-Sy...  
1  https://www.amazon.in/Ambrane-Unbreakable-Char...  
2  https://www.amazon.in/Sounce-iPhone-Charging-C...  
3  https://www.amazon.in/Deuce-300-Resistant-Tang...  
4  https://www.amazon.in/Portronics-Konnect-POR-1...

# Data Understanding

In [205]:
# Data Information
sales.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1465 entries, 0 to 1464
Data columns (total 16 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   product_id           1465 non-null   object
 1   product_name         1465 non-null   object
 2   category             1465 non-null   object
 3   discounted_price     1465 non-null   object
 4   actual_price         1465 non-null   object
 5   discount_percentage  1465 non-null   object
 6   rating               1465 non-null   object
 7   rating_count         1463 non-null   object
 8   about_product        1465 non-null   object
 9   user_id              1465 non-null   object
 10  user_name            1465 non-null   object
 11  review_id            1465 non-null   object
 12  review_title         1465 non-null   object
 13  review_content       1465 non-null   object
 14  img_link             1465 non-null   object
 15  product_link         1465 non-null   object
dtypes: obj

In [206]:
print(f'This dataset have {sales.shape[0]} Rows and {sales.shape[1]} Columns in total')

This dataset have 1465 Rows and 16 Columns in total


### Summary : 
* Dataset Shape is 1464 x 16
* Data Types are all object, expected columns to be numerics are 'discounted_price', 'actual_price', 'discount_percentage', 'rating', 'rating_count'
* There is some missing value on columns 'rating_count'

# Data Cleaning

## Change all the expected numerics columns/features from object

In [207]:
# Discounted Price
sales['discounted_price'] = sales['discounted_price'].replace('₹', '', regex=True).replace(',', '', regex=True).astype(float)

# Actual Price
sales['actual_price'] = sales['actual_price'].replace('₹', '', regex=True).replace(',', '', regex=True).astype(float)

# Discounted Percentage
sales['discount_percentage'] = sales['discount_percentage'].replace('%', '', regex=True).astype(float) / 100

# Rating Count
sales['rating_count'] = sales['rating_count'].replace(',','',regex=True).astype(float)

For the 'rating' we need to inspect a little more cause when we just convert to float it's give an error that there some non number value in the columns

In [208]:
sales['rating'].value_counts().reset_index()

rating  count
0     4.1    244
1     4.3    230
2     4.2    228
3     4.0    129
4     3.9    123
5     4.4    123
6     3.8     86
7     4.5     75
8       4     52
9     3.7     42
10    3.6     35
11    3.5     26
12    4.6     17
13    3.3     16
14    3.4     10
15    4.7      6
16    3.1      4
17    3.0      3
18    4.8      3
19    5.0      3
20    2.8      2
21    3.2      2
22    2.3      1
23      |      1
24      2      1
25      3      1
26    2.6      1
27    2.9      1

if we see from table given above, there's a '|' value on index 23

solution :
1. We can search the rating online and replace it with the rating we found
2. We can replace it with null and later impute it with median or mean

In [209]:
# Solution 2
sales['rating'] = sales['rating'].replace('|', np.nan).astype(float)

In [210]:
# New Data Information
sales.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1465 entries, 0 to 1464
Data columns (total 16 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   product_id           1465 non-null   object 
 1   product_name         1465 non-null   object 
 2   category             1465 non-null   object 
 3   discounted_price     1465 non-null   float64
 4   actual_price         1465 non-null   float64
 5   discount_percentage  1465 non-null   float64
 6   rating               1464 non-null   float64
 7   rating_count         1463 non-null   float64
 8   about_product        1465 non-null   object 
 9   user_id              1465 non-null   object 
 10  user_name            1465 non-null   object 
 11  review_id            1465 non-null   object 
 12  review_title         1465 non-null   object 
 13  review_content       1465 non-null   object 
 14  img_link             1465 non-null   object 
 15  product_link         1465 non-null   o

## Handling Missing Value

In [211]:
sales.isna().sum()

product_id             0
product_name           0
category               0
discounted_price       0
actual_price           0
discount_percentage    0
rating                 1
rating_count           2
about_product          0
user_id                0
user_name              0
review_id              0
review_title           0
review_content         0
img_link               0
product_link           0
dtype: int64

From the data given, we know that we have 3 missing data (2 rating_count and 1 rating)

### Rating

In [212]:
# Rating missing value data
sales.query('rating.isna()')

product_id                                       product_name  \
1279  B08L12N5H1  Eureka Forbes car Vac 100 Watts Powerful Sucti...   

                                               category  discounted_price  \
1279  Home&Kitchen|Kitchen&HomeAppliances|Vacuum,Cle...            2099.0   

      actual_price  discount_percentage  rating  rating_count  \
1279        2499.0                 0.16     NaN         992.0   

                                          about_product  \
1279  No Installation is provided for this product|1...   

                                                user_id  \
1279  AGTDSNT2FKVYEPDPXAA673AIS44A,AER2XFSWNN4LAUCJ5...   

                                              user_name  \
1279  Divya,Dr Nefario,Deekshith,Preeti,Prasanth R,P...   

                                              review_id  \
1279  R2KKTKM4M9RDVJ,R1O692MZOBTE79,R2WRSEWL56SOS4,R...   

                                           review_title  \
1279  Decent product,doesn't pick up sand,Ok ok,Must...   

                                         review_content  \
1279  Does the job well,doesn't work on sand. though...   

                                               img_link  \
1279  https://m.media-amazon.com/images/W/WEBP_40237...   

                                           product_link  
1279  https://www.amazon.in/Eureka-Forbes-Vacuum-Cle...

In [221]:
# Rating Information
r_median = sales['rating'].median()
r_mean = sales['rating'].mean()

print(f'Rating Median : {r_median}')
print(f'Rating Mean : {r_mean}')

Rating Median : 4.1
Rating Mean : 4.096584699453552


In [223]:
# Fill the missing value with median
sales['rating'] = sales['rating'].fillna(value=r_median)

### Rating Count

In [175]:
# Rating Count missing value data
sales.query('rating_count.isna()')

product_id                                       product_name  \
282  B0B94JPY2N  Amazon Brand - Solimo 65W Fast Charging Braide...   
324  B0BQRJ3C47  REDTECH USB-C to Lightning Cable 3.3FT, [Apple...   

                                              category  discounted_price  \
282  Computers&Accessories|Accessories&Peripherals|...             199.0   
324  Computers&Accessories|Accessories&Peripherals|...             249.0   

     actual_price  discount_percentage  rating  rating_count  \
282         999.0                 0.80     3.0           NaN   
324         999.0                 0.75     5.0           NaN   

                                         about_product  \
282  USB C to C Cable: This cable has type C connec...   
324  💎[The Fastest Charge] - This iPhone USB C cabl...   

                          user_id    user_name      review_id  \
282  AE7CFHY23VAJT2FI4NZKKP6GS2UQ       Pranav   RUB7U91HVZ30   
324  AGJC5O5H5BBXWUV7WRIEIOOR3TVQ  Abdul Gafur  RQXD5SAMMPC6L   

                                     review_title  \
282  The cable works but is not 65W as advertised   
324                               Awesome Product   

                                        review_content  \
282  I have a pd supported car charger and I bought...   
324  Quick delivery.Awesome ProductPacking was good...   

                                              img_link  \
282  https://m.media-amazon.com/images/W/WEBP_40237...   
324  https://m.media-amazon.com/images/I/31-q0xhaTA...   

                                          product_link  
282  https://www.amazon.in/Amazon-Brand-Charging-Su...  
324  https://www.amazon.in/REDTECH-Lightning-Certif...

In [225]:
# Rating Count Information
rc_median = sales['rating_count'].median()
rc_mean = sales['rating_count'].mean()

print(f'Rating Count Median : {rc_median}')
print(f'Rating Count Mean : {rc_mean}')

Rating Count Median : 5179.0
Rating Count Mean : 18295.541353383458


In [226]:
# Fill the missing data using mean
sales.rating_count = sales.rating_count.fillna(value=rc_median)

### Result

In [227]:
sales.isna().sum()

product_id             0
product_name           0
category               0
discounted_price       0
actual_price           0
discount_percentage    0
rating                 0
rating_count           0
about_product          0
user_id                0
user_name              0
review_id              0
review_title           0
review_content         0
img_link               0
product_link           0
dtype: int64

Missing value problem solved !

## Checking for duplicate data

In [229]:
# Checking for duplicate rows
print(sales.duplicated().any())

False


In [230]:
# Checking for duplicate data each columns/features
for i in sales.columns:
    print(f'Duplicate in {i.ljust(23)}: {sales[i].duplicated().sum()}')

Duplicate in product_id             : 114
Duplicate in product_name           : 128
Duplicate in category               : 1254
Duplicate in discounted_price       : 915
Duplicate in actual_price           : 1016
Duplicate in discount_percentage    : 1373
Duplicate in rating                 : 1440
Duplicate in rating_count           : 322
Duplicate in about_product          : 172
Duplicate in user_id                : 271
Duplicate in user_name              : 271
Duplicate in review_id              : 271
Duplicate in review_title           : 271
Duplicate in review_content         : 253
Duplicate in img_link               : 53
Duplicate in product_link           : 0


Key Point:
* There are some duplicates value in the 'product_id' columns
* Need to check other duplicate columns


In [231]:
# Row before removing all the dupes
rows_before = len(sales)

### Product ID

In [232]:
dupe_id = sales['product_id'].value_counts().reset_index()
dupe_id = dupe_id.query('count > 1')
dupe_id

product_id  count
0   B08WRWPM22      3
1   B01GGKYKQM      3
2   B09CMP1SC8      3
3   B09W5XR9RT      3
4   B09YLXYP7Y      3
5   B09KLVMZ3B      3
6   B08R69VDHT      3
7   B08CF3B7N1      3
8   B098NS6PVG      3
9   B082T6V3DT      3
10  B08Y1TFSP6      3
11  B08DDRGWTJ      3
12  B08CF3D7QR      3
13  B085DTN6R2      3
14  B07JW9H4J1      3
15  B083342NKJ      3
16  B09C6HXFC1      3
17  B09NHVCHS9      3
18  B077Z65HSD      3
19  B07XLCFSSN      3
20  B08HDJ86NZ      3
21  B096MSW6CT      3
22  B09YV4MW2T      2
23  B0B3RRWSF6      2
24  B09Z6WH2N1      2
25  B097R25DP7      2
26  B07GVGTSLN      2
27  B09MQSCJQ1      2
28  B094YFFSMY      2
29  B09MT84WV5      2
30  B0BMXMLSMM      2
31  B078G6ZF5Z      2
32  B0141EZMAI      2
33  B07RD611Z8      2
34  B0BDRVFDKP      2
35  B0B5LVS732      2
36  B0B9BXKBC7      2
37  B09V12K8NT      2
38  B01DEWVZ2C      2
39  B09C6HWG18      2
40  B00NH11KIK      2
41  B0B3N7LR6K      2
42  B09ZQK9X8G      2
43  B07KRCW6LZ      2
44  B096VF5YYF      2
45  B0B5ZF3NRK      2
46  B002PD61Y4      2
47  B07WG8PDCW      2
48  B0949SBKMP      2
49  B07232M876      2
50  B07P681N66      2
51  B08K4PSZ3V      2
52  B0994GFWBH      2
53  B01FSYQ2A4      2
54  B0BFWGBX61      2
55  B0B5B6PQCT      2
56  B08MTCKDYN      2
57  B085HY1DGR      2
58  B0B4DT8MKT      2
59  B08CDKQ8T6      2
60  B0B9BD2YL4      2
61  B08BCKN299      2
62  B01GGKZ0V6      2
63  B09F9YQQ7B      2
64  B082LZGK39      2
65  B09NVPSCQT      2
66  B0789LZTCJ      2
67  B07KSMBL2H      2
68  B09Q8HMKZX      2
69  B09YV4RG4D      2
70  B08BQ947H3      2
71  B082LSVT4B      2
72  B08WRBG3XW      2
73  B08DPLCM6T      2
74  B0B86CDHL1      2
75  B085194JFL      2
76  B09F6S8BT6      2
77  B0B3MWYCHQ      2
78  B01M4GGIVU      2
79  B094JNXNPV      2
80  B07XJYYH7L      2
81  B09RWZRCP1      2
82  B00NH11PEY      2
83  B09CMM3VGK      2
84  B08QSC1XY8      2
85  B008FWZGSG      2
86  B0B4HJNPV4      2
87  B08Y1SJVV5      2
88  B09KGV7WSV      2
89  B09RZS1NQT      2
90  B0B2DJDCPX      2
91  B07DJLFMPS      2

We have 92 'product_id' that have dupe data

In [ ]:
new_sample = sales.query('product_id == "B08WRWPM22"').reset_index()
new_sample

index  product_id                                       product_name  \
0      6  B08WRWPM22  boAt Micro USB 55 Tangle-free, Sturdy Micro US...   
1    418  B08WRWPM22  boAt Micro USB 55 Tangle-free, Sturdy Micro US...   
2    658  B08WRWPM22  boAt Micro USB 55 Tangle-free, Sturdy Micro US...   

                                            category  discounted_price  \
0  Computers&Accessories|Accessories&Peripherals|...            176.63   
1  Computers&Accessories|Accessories&Peripherals|...            176.63   
2  Computers&Accessories|Accessories&Peripherals|...            176.63   

   actual_price  discount_percentage  rating  rating_count  \
0         499.0                 0.65     4.1       15188.0   
1         499.0                 0.65     4.1       15189.0   
2         499.0                 0.65     4.1       15189.0   

                                       about_product  \
0  It Ensures High Speed Transmission And Chargin...   
1  It Ensures High Speed Transmission And Chargin...   
2  It Ensures High Speed Transmission And Chargin...   

                                             user_id  \
0  AG7C6DAADCTRQJG2BRS3RIKDT52Q,AFU7BOMPVJ7Q3TTA4...   
1  AG7C6DAADCTRQJG2BRS3RIKDT52Q,AFU7BOMPVJ7Q3TTA4...   
2  AG7C6DAADCTRQJG2BRS3RIKDT52Q,AFU7BOMPVJ7Q3TTA4...   

                                           user_name  \
0  Vivek kumar,Amazon Customer,SARTHAK,Chiranjeev...   
1  Vivek kumar,Amazon Customer,SARTHAK,Chiranjeev...   
2  Vivek kumar,Amazon Customer,SARTHAK,Chiranjeev...   

                                           review_id  \
0  R8E73K2KWJRDS,RSD0JTIIWQQL8,R64CRSTE9SLW1,R2FR...   
1  R8E73K2KWJRDS,RSD0JTIIWQQL8,R64CRSTE9SLW1,R2FR...   
2  R8E73K2KWJRDS,RSD0JTIIWQQL8,R64CRSTE9SLW1,R2FR...   

                                        review_title  \
0  Long durable.,good,Does not charge Lenovo m8 t...   
1  Long durable.,good,Does not charge Lenovo m8 t...   
2  Long durable.,good,Does not charge Lenovo m8 t...   

                                      review_content  \
0  Build quality is good and it is comes with 2 y...   
1  Build quality is good and it is comes with 2 y...   
2  Build quality is good and it is comes with 2 y...   

                                            img_link  \
0  https://m.media-amazon.com/images/W/WEBP_40237...   
1  https://m.media-amazon.com/images/I/41jlwEZpa5...   
2  https://m.media-amazon.com/images/W/WEBP_40237...   

                                        product_link  
0  https://www.amazon.in/boAt-Micro-USB-Tangle-Fr...  
1  https://www.amazon.in/boAt-Micro-USB-Tangle-Fr...  
2  https://www.amazon.in/boAt-Micro-USB-Tangle-Fr...

In [ ]:
# Check the detail in each columns for the dupes data
for i in new_sample.columns:
    print(f'Dupe in {i.ljust(23)}: {new_sample[i].duplicated().sum()}')

Dupe in index                  : 0
Dupe in product_id             : 2
Dupe in product_name           : 2
Dupe in category               : 2
Dupe in discounted_price       : 2
Dupe in actual_price           : 2
Dupe in discount_percentage    : 2
Dupe in rating                 : 2
Dupe in rating_count           : 1
Dupe in about_product          : 2
Dupe in user_id                : 2
Dupe in user_name              : 2
Dupe in review_id              : 2
Dupe in review_title           : 2
Dupe in review_content         : 2
Dupe in img_link               : 1
Dupe in product_link           : 0


it's weird that the product have the same id but have 1 different rating count and img link, but theres no dupe in product link 

In [ ]:
# Rating Count
print(f'{new_sample['rating_count']}')

   rating_count                                           img_link  \
0       15188.0  https://m.media-amazon.com/images/W/WEBP_40237...   
1       15189.0  https://m.media-amazon.com/images/I/41jlwEZpa5...   
2       15189.0  https://m.media-amazon.com/images/W/WEBP_40237...   

                                        product_link  
0  https://www.amazon.in/boAt-Micro-USB-Tangle-Fr...  
1  https://www.amazon.in/boAt-Micro-USB-Tangle-Fr...  
2  https://www.amazon.in/boAt-Micro-USB-Tangle-Fr...  


In [ ]:
# Looking into these weird dupes data
col = ['rating_count', 'img_link', 'product_link']

for i in col:
    print(i)
    for j in range(3):
        print(new_sample[i].iloc[j])
    print('\n')

rating_count
15188.0
15189.0
15189.0


img_link
https://m.media-amazon.com/images/W/WEBP_402378-T2/images/I/41jlwEZpa5L._SX300_SY300_QL70_FMwebp_.jpg
https://m.media-amazon.com/images/I/41jlwEZpa5L._SX300_SY300_QL70_ML2_.jpg
https://m.media-amazon.com/images/W/WEBP_402378-T2/images/I/41jlwEZpa5L._SX300_SY300_QL70_FMwebp_.jpg


product_link
https://www.amazon.in/boAt-Micro-USB-Tangle-Free-Transmission/dp/B08WRWPM22/ref=sr_1_8?qid=1672909124&s=electronics&sr=1-8
https://www.amazon.in/boAt-Micro-USB-Tangle-Free-Transmission/dp/B08WRWPM22/ref=sr_1_87?qid=1672895770&s=electronics&sr=1-87
https://www.amazon.in/boAt-Micro-USB-Tangle-Free-Transmission/dp/B08WRWPM22/ref=sr_1_78?qid=1672902998&s=computers&sr=1-78




In [280]:
for i in new_sample['rating_count'], new_sample['img_link']:
    print(i)

0    15188.0
1    15189.0
2    15189.0
Name: rating_count, dtype: float64
0    https://m.media-amazon.com/images/W/WEBP_402378-T2/images/I/41jlwEZpa5L._SX300_SY300_QL70_FMwebp_.jpg
1                               https://m.media-amazon.com/images/I/41jlwEZpa5L._SX300_SY300_QL70_ML2_.jpg
2    https://m.media-amazon.com/images/W/WEBP_402378-T2/images/I/41jlwEZpa5L._SX300_SY300_QL70_FMwebp_.jpg
Name: img_link, dtype: object


In [266]:
col = ['rating_count', 'img_link', 'product_link']

for i in col:
    print(f"\n{i}")  # Nama kolom
    print("\n".join(map(str, new_sample[i].iloc[:3])))  # Ambil 3 baris pertama



rating_count
15188.0
15189.0
15189.0

img_link
https://m.media-amazon.com/images/W/WEBP_402378-T2/images/I/41jlwEZpa5L._SX300_SY300_QL70_FMwebp_.jpg
https://m.media-amazon.com/images/I/41jlwEZpa5L._SX300_SY300_QL70_ML2_.jpg
https://m.media-amazon.com/images/W/WEBP_402378-T2/images/I/41jlwEZpa5L._SX300_SY300_QL70_FMwebp_.jpg

product_link
https://www.amazon.in/boAt-Micro-USB-Tangle-Free-Transmission/dp/B08WRWPM22/ref=sr_1_8?qid=1672909124&s=electronics&sr=1-8
https://www.amazon.in/boAt-Micro-USB-Tangle-Free-Transmission/dp/B08WRWPM22/ref=sr_1_87?qid=1672895770&s=electronics&sr=1-87
https://www.amazon.in/boAt-Micro-USB-Tangle-Free-Transmission/dp/B08WRWPM22/ref=sr_1_78?qid=1672902998&s=computers&sr=1-78


In [146]:
# Remove all the 'product_id' dupes from dataset
sales = sales.drop_duplicates(subset='product_id', keep='first')

# Rows after removing the dupes
rows_after_id = len(sales)


In [147]:
print(f'Rows before removing dupes id : {rows_before}')
print(f'Rows after removing dupes id : {rows_after_id}')
print(f'Succesfully removes {rows_before - rows_after_id} rows')

Rows before removing dupes id : 1465
Rows after removing dupes id : 1351
Succesfully removes 114 rows


### Product Name

In [148]:
dupe_name = sales['product_name'].value_counts().reset_index()
dupe_name = dupe_name.query('count > 1')
dupe_name

product_name  count
0  Fire-Boltt Ninja Call Pro Plus 1.83" Smart Wat...      5
1  Fire-Boltt Phoenix Smart Watch with Bluetooth ...      3
2  Samsung Galaxy M13 5G (Aqua Green, 6GB, 128GB ...      3
3  Redmi Note 11 (Horizon Blue, 6GB RAM, 64GB Sto...      2
4  Smashtronics® - Case for Firetv Remote, Fire S...      2
5  Fire-Boltt Visionary 1.78" AMOLED Bluetooth Ca...      2
6  Fire-Boltt India's No 1 Smartwatch Brand Talk ...      2
7  Samsung Galaxy M33 5G (Emerald Brown, 6GB, 128...      2
8                                   Firestick Remote      2

After removing all the dupe id, we have 8 remaining product_name who have dupes also

In [149]:
name = dupe_name['product_name'].iloc[0]
name

'Fire-Boltt Ninja Call Pro Plus 1.83" Smart Watch with Bluetooth Calling, AI Voice Assistance, 100 Sports Modes IP67 Rating, 240*280 Pixel High Resolution'

In [150]:
sales[sales['product_name'] == name]

product_id                                       product_name  \
334  B0BF57RN3K  Fire-Boltt Ninja Call Pro Plus 1.83" Smart Wat...   
364  B0BF54972T  Fire-Boltt Ninja Call Pro Plus 1.83" Smart Wat...   
368  B0BF563HB4  Fire-Boltt Ninja Call Pro Plus 1.83" Smart Wat...   
372  B0BF4YBLPX  Fire-Boltt Ninja Call Pro Plus 1.83" Smart Wat...   
380  B0BF54LXW6  Fire-Boltt Ninja Call Pro Plus 1.83" Smart Wat...   

                                        category  discounted_price  \
334  Electronics|WearableTechnology|SmartWatches            1799.0   
364  Electronics|WearableTechnology|SmartWatches            1799.0   
368  Electronics|WearableTechnology|SmartWatches            1799.0   
372  Electronics|WearableTechnology|SmartWatches            1799.0   
380  Electronics|WearableTechnology|SmartWatches            1799.0   

     actual_price  discount_percentage  rating  rating_count  \
334       19999.0                 91.0     4.2       13937.0   
364       19999.0                 91.0     4.2       13937.0   
368       19999.0                 91.0     4.2       13937.0   
372       19999.0                 91.0     4.2       13937.0   
380       19999.0                 91.0     4.2       13937.0   

                                         about_product  \
334  Fire-Boltt is India' No 1 Wearable Watch Brand...   
364  Fire-Boltt is India' No 1 Wearable Watch Brand...   
368  Fire-Boltt is India' No 1 Wearable Watch Brand...   
372  Fire-Boltt is India' No 1 Wearable Watch Brand...   
380  Fire-Boltt is India' No 1 Wearable Watch Brand...   

                                               user_id  \
334  AEC6UDCEAUIBIFHGQDQ4KR67GC4A,AHRKSUOZXKKDERRY3...   
364  AEC6UDCEAUIBIFHGQDQ4KR67GC4A,AHRKSUOZXKKDERRY3...   
368  AEC6UDCEAUIBIFHGQDQ4KR67GC4A,AHRKSUOZXKKDERRY3...   
372  AEC6UDCEAUIBIFHGQDQ4KR67GC4A,AHRKSUOZXKKDERRY3...   
380  AEC6UDCEAUIBIFHGQDQ4KR67GC4A,AHRKSUOZXKKDERRY3...   

                                             user_name  \
334  venkatesh kg,Raghuram,YOGESH,Manan Parmar,Niku...   
364  venkatesh kg,Raghuram,YOGESH,Manan Parmar,Niku...   
368  venkatesh kg,Raghuram,YOGESH,Manan Parmar,Niku...   
372  venkatesh kg,Raghuram,YOGESH,Manan Parmar,Niku...   
380  venkatesh kg,Raghuram,YOGESH,Manan Parmar,Niku...   

                                             review_id  \
334  R1PKIMKR1E8X8T,R23UV7ZBIEEZD3,RYRHNVDKS5RFY,RS...   
364  R1PKIMKR1E8X8T,R23UV7ZBIEEZD3,RYRHNVDKS5RFY,RS...   
368  R1PKIMKR1E8X8T,R23UV7ZBIEEZD3,RYRHNVDKS5RFY,RS...   
372  R1PKIMKR1E8X8T,R23UV7ZBIEEZD3,RYRHNVDKS5RFY,RS...   
380  R1PKIMKR1E8X8T,R23UV7ZBIEEZD3,RYRHNVDKS5RFY,RS...   

                                          review_title  \
334  Worth the money,Good smart watch,the screen do...   
364  Worth the money,Good smart watch,the screen do...   
368  Worth the money,Good smart watch,the screen do...   
372  Worth the money,Good smart watch,the screen do...   
380  Worth the money,Good smart watch,the screen do...   

                                        review_content  \
334  I liked it, worth the money, light weight and ...   
364  I liked it, worth the money, light weight and ...   
368  I liked it, worth the money, light weight and ...   
372  I liked it, worth the money, light weight and ...   
380  I liked it, worth the money, light weight and ...   

                                              img_link  \
334  https://m.media-amazon.com/images/I/41WCgGbvwh...   
364  https://m.media-amazon.com/images/I/41mzbWC6Ak...   
368  https://m.media-amazon.com/images/I/41ivjqdXb0...   
372  https://m.media-amazon.com/images/I/410VGCE+q2...   
380  https://m.media-amazon.com/images/I/41Fq27ZjJf...   

                                          product_link  
334  https://www.amazon.in/Fire-Boltt-Bluetooth-Cal...  
364  https://www.amazon.in/Fire-Boltt-Bluetooth-Cal...  
368  https://www.amazon.in/Fire-Boltt-Bluetooth-Cal...  
372  https://www.amazon.in/Fire-Boltt-Bluetooth-Cal...  
380  https://www.amazon.in/Fire-Boltt-Bluetooth-Ca

the data shown only have different id but all the remaining features are the same. So we need to erase them also

In [151]:
# Remove all the 'product_name' dupes from dataset
sales = sales.drop_duplicates(subset='product_name',keep='first')

# Rows after removing name dupes
rows_after_name = len(sales)

In [152]:
print(f'Rows before removing name dupes : {rows_after_id}')
print(f'Rows after removing name dupes : {rows_after_name}')
print(f'Succesfully removes {rows_after_id - rows_after_name} rows')

Rows before removing name dupes : 1351
Rows after removing name dupes : 1337
Succesfully removes 14 rows


### Duplicate re-check

In [154]:
# Re-check the number of duplicate in product_id and product_name
for i in sales.columns:
    print(f'Duplicates in {i.ljust(23)}: {sales[i].duplicated().sum()}')

Duplicates in product_id             : 0
Duplicates in product_name           : 0
Duplicates in category               : 1126
Duplicates in discounted_price       : 788
Duplicates in actual_price           : 888
Duplicates in discount_percentage    : 1245
Duplicates in rating                 : 1312
Duplicates in rating_count           : 222
Duplicates in about_product          : 53
Duplicates in user_id                : 152
Duplicates in user_name              : 152
Duplicates in review_id              : 152
Duplicates in review_title           : 152
Duplicates in review_content         : 141
Duplicates in img_link               : 21
Duplicates in product_link           : 0
